elyra/contents/file_parser.py

```
@@ -0,0 +1,193 @@
+        filepath = kwargs['filepath']
+
+        if '.ipynb' in filepath:
+            return NotebookFileParser(filepath)
+        elif '.py' in filepath:
+            return PythonFileParser(filepath)
+        elif '.r' in filepath:
+            return RFileParser(filepath)
+        else:
+            raise ValueError('Unsupported file type: {}'.format(filepath))
+
+    _type = None
+
+    def __init__(self, operation_filepath):
+        self._operation_filepath = operation_filepath
+        self.parser = None
+
+        # Consider removing or add log messages
+        self.log = log.get_logger()
+
+    @property
+    @abstractmethod
+    def type(self):
+        pass
+
+    @property
+    def operation_filepath(self):
+        # Retrieve the absolute path
+        root_dir = get_expanded_path()
+        abs_path = get_absolute_path(root_dir, self._operation_filepath)
+
+        if not os.path.exists(abs_path):
+            raise FileNotFoundError(f'Could not find {abs_path}')
+        if not os.path.isfile(abs_path):
+            raise ValueError(f'Not a file: {abs_path}')
+
+        self._operation_filepath = abs_path
+        return self._operation_filepath
+
+    @abstractmethod
+    def get_next_code_chunk(self) -> List[str]:
+        # Implements a generator for lines of code for the specified operation type
+        with open(self.operation_filepath) as f:
+            for line in f:
+                yield [line.strip()]
+
+    def get_resources(self):
+        # Returns a model dictionary of all the regex matches for each key in the regex dictionary
+        model = dict()
+        model["env_list"] = dict()
+        model["inputs"] = dict()
+        model["outputs"] = dict()
+
+        for chunk in self.get_next_code_chunk():
+            if chunk:
+                for line in chunk:
+                    if not self.parser:
+                        raise RuntimeError(f'Could not find appropriate language parser for {self.operation_filepath}')
```

This should be moved, to the first thing this method checks.  Or, perhaps better, make `parser` a property, since it should have a value following construction - then let the accessor throw a `ValueError` if it finds it not set.

---

Ugh, of course! No idea why I put it there yesterday ha